tests/ibr/test_service.py

```diff
@@ -0,0 +1,307 @@
+
+        _, (iq, ), _ = self.cc.send.mock_calls[0]
+
+        self.assertIsInstance(
+            iq,
+            aioxmpp.IQ,
+        )
+
+        self.assertEqual(
+            iq.type_,
+            aioxmpp.IQType.SET,
+        )
+
+        self.assertEqual(
+            iq.to,
+            TEST_PEER.bare().replace(localpart=None),
+        )
+
+        self.assertIsInstance(
+            iq.payload,
+            ibr_xso.Query,
+        )
+
+        self.assertEqual(
+            iq.payload.username,
+            TEST_PEER.localpart,
+        )
+
+        self.assertEqual(
+            iq.payload.password,
+            new_pass,
+        )
+
+        if aux_fields is not None:
+            for key, value in aux_fields.items():
+                self.assertEqual(
+                    getattr(iq.payload, key),
+                    value,
+                )
+
+    def test_ibr_cancel_registration(self):
+        self.cc.send.return_value = None
+        self.cc.local_jid = TEST_PEER
+
+        run_coroutine(self.s.cancel_registration(self.cc))
+
+        self.cc.send.assert_called_once_with(unittest.mock.ANY)
+
+        _, (iq, ), _ = self.cc.send.mock_calls[0]
+
+        self.assertIsInstance(
+            iq,
+            aioxmpp.IQ,
+        )
+
+        self.assertEqual(
+            iq.to,
+            TEST_PEER.bare().replace(localpart=None),
+        )
+
+        self.assertEqual(
+            iq.type_,
+            aioxmpp.IQType.SET,
+        )
+
+        self.assertIsInstance(
+            iq.payload,
+            ibr_xso.Query,
+        )
+
+        self.assertEqual(
+            iq.payload.remove,
+            True,
+        )
+
+    def test_ibr_get_registration_fields(self):
+        with unittest.mock.patch(
+                'aioxmpp.ibr.RegistrationService.connect_and_send'
+        ) as mock1:
+            mock1.return_value = self.auxiliar2()
+            res = run_coroutine(self.s.get_registration_fields(TEST_PEER))
+
+            _, (iq, *_), _ = mock1.mock_calls[0]
+
+            self.assertIsInstance(
+                iq,
+                aioxmpp.IQ,
+            )
+
+            self.assertEqual(
+                iq.to,
+                TEST_PEER.bare().replace(localpart=None),
+            )
+
+            self.assertEqual(
+                iq.type_,
+                aioxmpp.IQType.GET,
+            )
+
+            self.assertIsInstance(
+                iq.payload,
+                ibr_xso.Query,
+            )
+
+            self.assertIsInstance(
+                res,
+                list,
+            )
+
+            self.assertEqual(
+                res,
+                ['username'],
+            )
+
+    def test_ibr_register(self):
+        password = "aaa"
+        aux_fields = {"nick": "romeo's lover"}
+        with unittest.mock.patch(
+                'aioxmpp.ibr.RegistrationService.connect_and_send'
+        ) as mock1:
+            mock1.return_value = self.auxiliar3()
```

Comment:
I think you can use a CoroutineMock from aioxmpp.testutils here. If you need an example of that, let me know; it basically operates like a normal mock, except that it emulates a coroutine. Usage with ``patch`` is like: ``unittest.mock.patch(…, new=CoroutineMock())``, otherwise it’s pretty much the same.